File: ros2_project/src/2-3ActionCommunication/action_communication_cpp/src/action_client.cpp
#include <memory>

#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"

#include "action_interfaces/action/fibonacci.hpp"

using Fibonacci = action_interfaces::action::Fibonacci;
using GoalHandleFibonacci = rclcpp_action::ClientGoalHandle<Fibonacci>;

class FibonacciActionClient : public rclcpp::Node
{
public:
    // 构造函数
    FibonacciActionClient(): Node("fibonacci_action_client")
    {
        // 创建 Action 客户端
        this->client_ = rclcpp_action::create_client<Fibonacci>(this,"fibonacci");

        // 等待 Action 服务端上线
        // 如果 Action 服务端还没启动，这里会阻塞等待
        this->client_->wait_for_action_server();

        // 构建一个 Goal 消息
        // Fibonacci::Goal 对应 .action 文件中的 Goal 部分
        Fibonacci::Goal goal_msg;
        goal_msg.order = 10;

        // 异步发送 Goal（即 目标传递服务 中 客户端发送请求给服务端）
        rclcpp_action::Client<Fibonacci>::SendGoalOptions options;
        options.feedback_callback =std::bind(&FibonacciActionClient::feedback_callback,this,std::placeholders::_1,std::placeholders::_2);
        options.result_callback =std::bind(&FibonacciActionClient::result_callback,this,std::placeholders::_1);
        this->client_->async_send_goal(goal_msg, options);
    }

private:
    // 反馈话题 的回调函数
    // 当服务端在 execute_callback 中调用 publish_feedback 时触发
    void feedback_callback(GoalHandleFibonacci::SharedPtr goal_handle,const std::shared_ptr<const Fibonacci::Feedback> feedback)
    {
        RCLCPP_INFO(this->get_logger(),"接收服务端反馈: [%d, %d, %d ...]",feedback->partial_sequence[0],feedback->partial_sequence[1],feedback->partial_sequence.back());
    }

    // 结果传递服务 中 客户端接收服务端数据的回调函数
    void result_callback(const GoalHandleFibonacci::WrappedResult & result)
    {
        RCLCPP_INFO(this->get_logger(), "结果:");
        for (auto number : result.result->sequence) 
        {
            RCLCPP_INFO(this->get_logger(), "%d", number);
        }
    }

    // 声明一个Action服务端
    rclcpp_action::Client<Fibonacci>::SharedPtr client_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<FibonacciActionClient>());
  rclcpp::shutdown();
  return 0;
}

File: ros2_project/src/2-3ActionCommunication/action_communication_py/action_communication_py/action_client.py
import rclpy
from rclpy.node import Node
from rclpy.action import ActionClient

from action_interfaces.action import Fibonacci

class FibonacciActionClient(Node):
    # 构造函数
    def __init__(self):
        super().__init__('fibonacci_action_client')

        # 创建Action客户端
        self._client = ActionClient(self, Fibonacci, 'fibonacci')

        # 等待 Action 服务端上线
        # 如果Action服务端还没启动，这里会阻塞等待
        self._client.wait_for_server()

        # 构建一个Goal消息
        # Fibonacci.Goal 对应 .action 文件中的 Goal 部分
        goal_msg = Fibonacci.Goal()
        goal_msg.order = 10
        
        # 异步发送 Goal（即 目标传递服务 中 客户端发送请求给服务端）
        self._send_goal_future = self._client.send_goal_async(
            goal_msg,                                           # 请求的内容
            feedback_callback=self.feedback_callback            # 注册反馈回调函数（反馈话题的回调函数）
        )

        # 注册 目标传递服务 响应回调函数
        # 当 Action 服务端对 Goal 作出“是否接受”的决定后，该 future 完成，并自动调用 goal_response_callback
        self._send_goal_future.add_done_callback(self.goal_response_callback)

    # 反馈话题 的回调函数
    # 当服务端在 execute_callback 中调用 publish_feedback 时触发
    def feedback_callback(self, feedback_msg):
        self.get_logger().info(f'接受服务端反馈: {feedback_msg.feedback.partial_sequence}')

    # 目标传递服务 中 客户端接收服务端数据的回调函数
    def goal_response_callback(self, future):
        # 获取futre中的内容
        goal_handle = future.result()
        if not goal_handle.accepted:
            self.get_logger().info('目标被拒绝！')
            return
        else:
            self.get_logger().info('目标被接受！')

            # 结果传递服务（Get Result）
            # 向 结果传递服务 发送请求，请求该 Goal 的最终执行结果
            self._get_result_future = goal_handle.get_result_async()                # 向 结果传递服务 发出请求（异步调用），get_result_async函数会返回一个future对象
            self._get_result_future.add_done_callback(self.get_result_callback)     # 注册 结果传递服务 的响应回调函数。当服务端处理完毕后的结果返回时，自动调用 get_result_callback函数

    # 结果传递服务 中 客户端接收服务端数据的回调函数
    def get_result_callback(self, future):
        result = future.result().result
        self.get_logger().info(f'结果: {result.sequence}')


def main():
    rclpy.init()
    node = FibonacciActionClient()
    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == '__main__':
    main()
